### ISSUE TITLE
Difficulty connecting to Google IoT Core

### ISSUE THREAD

**Issue by Radagan** [redboltz/mqtt_cpp#343]
I'm trying to connect to Google IoT Core which uses TLS 1.2 and requires it's root certificates. I'm setting those by filename with:

`client->set_ca_cert_file(roots);`

where roots is "roots.pem" and is the filename for a certificate file in pem format downloaded from Google. The file is in the current directory when the code is run. 

Has anyone had luck connecting to Google IoT with mqtt_cpp?

**Comment by jonesmz** [comment]
Have you called

set_default_verify_paths()

or

set_verify_mode()

or the other functions,

to set the SSL behavior?

**Comment by Radagan** [comment]
No, I haven't seen those. I followed the "mqtt_cpp/example/tls_client.cpp" example code and it only calls:

set_ca_cert_file()

I will have a look at those two methods. Thank you. 

The roots.pem file Google expects contains a number of certificates, it's in the current directory, and I pass simply the filename to set_ca_cert_file(). 

I'm very excited about this library and I'm looking forward to completely replacing paho.mqtt.cpp with it. Thank you for your help and patience.

**Comment by jonesmz** [comment]
I suspect what you might be running into is that the Google IOT service is using TLS 1.2, and the Boost ASIO defaults to a different version of TLS, so it's getting grumpy.

You can also set callbacks to the ssl context, and the on_error functions for mqtt_cpp, which can give you more information about why it's failing.

**Comment by Radagan** [comment]
That's a good point. Google does require TLS 1.2. I'll try again with some callbacks set.

**Comment by Radagan** [comment]
Actually, no... But I got a little time to work on this and decided to take just the tls example code and add the jwt token generation and other Google IoT Core idiosyncrasies. It connected immediately and I saw all the callbacks. So, I'll close this issue since clearly it's not related to Google IoT. If I get clean working code, maybe I'll open a pull request to add a connect option for Google IoT Core if anyone is interested. 

Weirdly, my project code that fails uses almost exactly the same functional code as the tls example, but with the tls client as a member of the class. It doesn't so much fail, as I never see a single callback. The only difference in the setting of the callbacks is that I'm capturing "this" in the lambda for some of them, including connack_handler. Maybe I'll try dropping that and see if it's called next.
